File: app/characters/base/character.py
import pygame

class Character:
    def __init__(self, images, x, y, speed, gravity):
        self.images = images
        self.x = x
        self.y = y
        self.speed = speed
        self.gravity = gravity
        self.frame_index = 0
        self.animation_timer = 0
        self.on_ground = False
        self.current_image = self.images[0]
        self.width = self.current_image.get_width()
        self.height = self.current_image.get_height()
        self.facing_left = False

    def draw(self, screen, camera):
        current_image = self.images[self.frame_index]
        if self.facing_left:
            current_image = pygame.transform.flip(current_image, True, False)
        screen.blit(current_image, (self.x - camera.x, self.y - camera.y))

    def move(self, keys, map_instance, TILE_SIZE, clock):
        is_moving = False

        if keys[pygame.K_LEFT] and not map_instance.check_collision(
            self.x, self.y, self.width, self.height, "left", self.speed
        ):
            self.x -= self.speed
            self.facing_left = True
            is_moving = True

        if keys[pygame.K_RIGHT] and not map_instance.check_collision(
            self.x, self.y, self.width, self.height, "right", self.speed
        ):
            self.x += self.speed
            self.facing_left = False
            is_moving = True

        self.on_ground, new_y = map_instance.is_on_ground(self.x, self.y, self.height)
        if self.on_ground:
            self.y = new_y
        else:
            self.y += self.gravity

        if is_moving:
            self.animation_timer += clock.get_time()
            if self.animation_timer >= 60:
                self.animation_timer = 0
                self.frame_index = (self.frame_index + 1) % len(self.images)

File: frontend_src/node/frontend/src/Character.ts
import { Map } from "./map";

export class Character {
  public images: HTMLImageElement[];
  public x: number;
  public y: number;
  public width: number;
  public height: number;
  public speed: number;
  public gravity: number;
  public frameIndex: number;
  public animationTimer: number;
  public onGround: boolean;
  public currentImage: HTMLImageElement;
  public facingLeft: boolean;

  constructor(
    images: HTMLImageElement[],
    x: number,
    y: number,
    speed: number,
    gravity: number,
  ) {
    this.images = images;
    this.x = x;
    this.y = y;
    this.speed = speed;
    this.gravity = gravity;
    this.frameIndex = 0;
    this.animationTimer = 0;
    this.onGround = false;
    this.currentImage = this.images[0];
    this.facingLeft = false;
    this.width = this.currentImage.width
    this.height = this.currentImage.height
  }

  draw(ctx: CanvasRenderingContext2D, cameraX: number, cameraY: number, width?: number, height?: number): void {
    if (!ctx || !(ctx instanceof CanvasRenderingContext2D)) {
      console.error("Invalid canvas context provided.");
      return;
    }
  
    const currentImage = this.images[this.frameIndex];
  
    if (!currentImage.complete) {
      console.warn("Image not loaded yet:", currentImage.src);
      return;
    }

    const drawWidth = width || currentImage.width;
    const drawHeight = height || currentImage.height;
  
    if (this.facingLeft) {
      ctx.save();
      ctx.scale(-1, 1);
      ctx.drawImage(
        currentImage,
        -(this.x + drawWidth - cameraX),
        this.y - cameraY,
        drawWidth,
        drawHeight,
      );
      ctx.restore();
    } else {
      ctx.drawImage(
        currentImage,
        this.x - cameraX,
        this.y - cameraY,
        drawWidth,
        drawHeight,
      );
    }
  }
  
  move(
    keys: { [key: string]: boolean },
    mapInstance: Map
  ): void {
  
    let isMoving = false;
  
    // 左への移動
    if (
      keys["ArrowLeft"] &&
      !mapInstance.checkCollision(this.x, this.y, this.width, this.height, "left", this.speed)
    ) {
      this.x -= this.speed;
      this.facingLeft = true;
      isMoving = true;
    }
  
    // 右への移動
    if (
      keys["ArrowRight"] &&
      !mapInstance.checkCollision(this.x, this.y, this.width, this.height, "right", this.speed)
    ) {
      this.x += this.speed;
      this.facingLeft = false;
      isMoving = true;
    }
  
    // 接地判定
    const [onGround, newY] = mapInstance.isOnGround(this.x, this.y, this.height);
    this.onGround = onGround;
    if (this.onGround) {
      this.y = newY;
    } else {
      this.y += this.gravity;
    }
  
    // アニメーション更新
    if (isMoving) {
      this.animationTimer += 1;
      if (this.animationTimer >= 10) {
        this.animationTimer = 0;
        this.frameIndex = (this.frameIndex + 1) % this.images.length;
      }
    } else {
    }
  }

  updateAnimation(FPS: number, isMoving: boolean): void {
    if (isMoving) {
      this.animationTimer += 1;
      if (this.animationTimer >= FPS / 6) { // 6 FPS のアニメーション速度
        this.animationTimer = 0;
        this.frameIndex = (this.frameIndex + 1) % this.images.length;
      }
    } else {
      this.frameIndex = 0; // 動いていない場合は静止状態のフレームに戻す
    }
  }
}
